### ISSUE TITLE
Scopes on Recursive Regex Cause Problems 

### ISSUE THREAD

**Issue by jeff-hykin** [microsoft/vscode-textmate#208]

**Comment by RedCMD** [comment]
dup:
https://github.com/microsoft/vscode-textmate/issues/127
https://github.com/microsoft/vscode-textmate/issues/164

**Comment by slevithan** [comment]
@RedCMD its not currently clear that any of these 3 issues are duplicates.

@jeff-hykin TextMate grammars use the Oniguruma regex engine, and Oniguruma has particular, non-portable (and not particularly intuitive) behavior about how subroutines affect captures and backreferences (e.g. see the details I mentioned in #164). The concept of non-participating captures (and their specific handling in various cases) is also likely relevant, and doesn't work the same in every regex flavor.

So, are you saying that you wish Oniguruma worked differently? In which case this should probably be closed (and any potential solution to your problem should probably be reframed around actual Oniguruma results), or are you saying there's an issue with vscode-textmate and it's not giving you the results you'd expect compared to how Oniguruma works? This distinction is important and frames any potential solution.

**Comment by jeff-hykin** [comment]
I want to push back a bit on this being marked as an upstream oniguruma issue. If this was merely a bug in oniguruma, then yes I'd agree it should be reported there so basically all textmate engines could "get the fix".

The issue though is not a bug, but a design flaw. Meaning if oniguruma changed how it fundamentally handled capture groups I believe it would break downstream applications.

We can't merely push this issue onto oniguruma, because it would be irresponsible for a tool to make such a breaking change. Meaning the end application has to change tooling to fix the problem.

I'm sure I won't be taken up on the offer, but the short of it is: the only way grammars will ever be highlighted correctly in VS Code is if VS code makes a small fork of oniguruma, or switches regex engines.

**Comment by jeff-hykin** [comment]
I do want to point out, swapping engines might not be as insane as it initially sounds. 

Translating oniguruma regex to PCRE is straightforward, and I don't know of any (desireable) features that are unique to oniguruma.

Its the C api of Oniguruma VS PCRE that I'm unaware of, and where all the work would probably be.

**Comment by jeff-hykin** [comment]
TLDR; I expect this to be closed as "an issue, but too much work to fix" 

If the effort on this could be spent on getting the tree sitter working, I'd prefer that anyways.

**Comment by slevithan** [comment]
> swapping engines might not be as insane as it initially sounds.

Unfortunately, it is. There's a long tail of TM grammars that rely on specific quirks of Oniguruma syntax and behavior.

> Translating oniguruma regex to PCRE is straightforward

That's far from true when dealing with the long tail of TM grammars. This is an area I'm very knowledgeable about since I created [Oniguruma-To-ES](https://github.com/slevithan/oniguruma-to-es) (which is almost certainly the most complex+accurate regex flavor-to-flavor translator that exists). It accurately translates Oniguruma to JS `RegExp` for 99.9+% of regexes from a sample of tens of thousands found in the hundreds of TM grammars provided by [Shiki](https://shiki.style/), resulting in current support of 98% of grammars. But let me tell you, this was ***not*** straightforward.

Oniguruma to PCRE2 would be significantly easier than translating to JS, but only to a point. Of course, you're talking about swapping engines rather than translating flavors. But the reason translation is hard is because Oniguruma is ***loaded*** with syntax and behavior edge cases that are nonportable and defy prior art. Subroutines are used very heavily in TM grammars (and interact with various other Oniguruma features in complicated ways), so you can bet that some are relying on Oniguruma's (inferior, less intuitive, and more complicated) subroutine behavior.

**Comment by slevithan** [comment]
I think it's hard to analyze. It's true that many esoteric Oniguruma features/sub-features are rarely used (or not used at all) in popular TM grammars. But one or two grammars might use them in one regex, making the whole grammar reliant on it. And of the features that are used... *damn* there are a lot of subtle syntax and behavior edge cases to get right if you want identical highlighting results and error handling for all grammars and all potential target strings. And those behavioral edge cases can't be analyzed by just counting features used.

On the specific point of TM grammars relying on how subroutines replace the values in their capture slots: it's not many cases, but there were some, and I had to build a complex system using a JS `RegExp` subclass to transfer the captured values into the right capture slots on match results in order to support this in [Shiki](https://shiki.style/)'s JS RegExp engine. And there are also various other subroutine-related quirks that work differently from PCRE, including wildly complicated interactions between Oniguruma subroutines and features like backreference multiplexing (an Oniguruma-specific feature), duplicate group names (which subroutines can't reference directly but might reference indirectly), and flag modifiers.

**Comment by RedCMD** [comment]
Onigmo seems to have a disabled option just for this
https://github.com/textmate/Onigmo/blob/03411d106584594b606b1d30757ec560954e6c49/doc/RE#L499-L507
